app/controllers/ApplicationController.scala

```diff
@@ -48,63 +48,78 @@ class ApplicationController @Inject() (
+  def uplift(redirectUrl: Option[RedirectUrl]): Action[AnyContent] = Action.async {
+
+    redirectUrl.fold(Future.successful(Redirect(routes.HomeController.index.url))) {
+      _.getEither(OnlyRelative) match {
+        case Right(safeRedirectUrl) => Future.successful(Redirect(safeRedirectUrl.url))
+        case Left(error)            => Future.successful(BadRequest(error))
+      }
+    }
   }
 
-  def showUpliftJourneyOutcome(continueUrl: Option[SafeRedirectUrl]): Action[AnyContent] =
+  def showUpliftJourneyOutcome(continueUrl: Option[RedirectUrl]): Action[AnyContent] =
     Action.async { implicit request =>
-      val journeyId =
-        List(request.getQueryString("token"), request.getQueryString("journeyId")).flatten.headOption
-
-      val retryUrl = routes.ApplicationController.uplift(continueUrl).url
-
-      journeyId match {
-        case Some(jid) =>
-          identityVerificationFrontendService
-            .getIVJourneyStatus(jid)
-            .map {
-              case Success =>
-                Ok(successView(continueUrl.map(_.url).getOrElse(routes.HomeController.index.url)))
-
-              case InsufficientEvidence =>
-                Redirect(routes.SelfAssessmentController.ivExemptLandingPage(continueUrl))
-
-              case UserAborted =>
-                logErrorMessage(UserAborted.toString)
-                Unauthorized(cannotConfirmIdentityView(retryUrl))
-
-              case FailedMatching =>
-                logErrorMessage(FailedMatching.toString)
-                Unauthorized(cannotConfirmIdentityView(retryUrl))
-
-              case Incomplete =>
-                logErrorMessage(Incomplete.toString)
-                Unauthorized(failedIvIncompleteView(retryUrl))
-
-              case PrecondFailed =>
-                logErrorMessage(PrecondFailed.toString)
-                Unauthorized(cannotConfirmIdentityView(retryUrl))
-
-              case LockedOut =>
-                logErrorMessage(LockedOut.toString)
-                Unauthorized(lockedOutView())
-
-              case Timeout =>
-                logErrorMessage(Timeout.toString)
-                Unauthorized(timeOutView())
-
-              case TechnicalIssue =>
-                logger.warn(s"TechnicalIssue response from identityVerificationFrontendService")
-                InternalServerError(technicalIssuesView(retryUrl))
-
-              case _ =>
-                InternalServerError(technicalIssuesView(retryUrl))
-            }
-            .getOrElse(BadRequest(technicalIssuesView(retryUrl)))
-        case _         =>
-          logger.error("journeyId missing or incorect")
-          Future.successful(InternalServerError(technicalIssuesView(retryUrl)))
+      val safeUrl: Either[String, Option[SafeRedirectUrl]] = continueUrl match {
+        case None    => Right(None)
+        case Some(x) => x.getEither(OnlyRelative).map(Some.apply)
+      }
+
+      safeUrl match {
+        case Left(error)    => Future.successful(BadRequest(error))
```

Comment:
This is quite complex. Cyclomatic complexity 17. I'd separate into private method